File: SC/Assignment2.cpp
#include <iostream>   // for input/output
#include <vector>     // to use dynamic 2D arrays (matrices)
#include <algorithm>  // for max() and min() functions
using namespace std;


vector<vector<double>> fuzzyUnion(const vector<vector<double>>& R, const vector<vector<double>>& S) {
    // create a matrix of same size as R (and S)
    vector<vector<double>> result(R.size(), vector<double>(R[0].size()));
    
    // loop through rows
    for (size_t i = 0; i < R.size(); i++) {
        // loop through columns
        for (size_t j = 0; j < R[0].size(); j++) {
            // union rule: take maximum of membership values
            result[i][j] = max(R[i][j], S[i][j]);
        }
    }
    return result; // return the resulting matrix
}


vector<vector<double>> fuzzyIntersection(const vector<vector<double>>& R, const vector<vector<double>>& S) {
    // create result matrix with same size
    vector<vector<double>> result(R.size(), vector<double>(R[0].size()));
    
    // loop through all cells
    for (size_t i = 0; i < R.size(); i++) {
        for (size_t j = 0; j < R[0].size(); j++) {
            // intersection rule: take minimum of membership values
            result[i][j] = min(R[i][j], S[i][j]);
        }
    }
    return result;
}


vector<vector<double>> fuzzyComplement(const vector<vector<double>>& R) {
    // create result matrix of same size
    vector<vector<double>> result(R.size(), vector<double>(R[0].size()));
    
    for (size_t i = 0; i < R.size(); i++) {
        for (size_t j = 0; j < R[0].size(); j++) {
            // complement rule: subtract value from 1
            result[i][j] = 1.0 - R[i][j];
        }
    }
    return result;
}

// Function to calculate composition of two fuzzy relations (R ○ S)
vector<vector<double>> fuzzyComposition(const vector<vector<double>>& R, const vector<vector<double>>& S) {
    size_t m = R.size();       // number of rows in R
    size_t n = R[0].size();    // number of columns in R (also rows in S)
    size_t p = S[0].size();    // number of columns in S

    // initialize result matrix with zeros (m x p)
    vector<vector<double>> result(m, vector<double>(p, 0.0));

    // loop through rows of R
    for (size_t i = 0; i < m; i++) {
        // loop through columns of S
        for (size_t k = 0; k < p; k++) {
            double maxVal = 0.0; // store maximum(min(R[i][j], S[j][k]))
            
            // loop through intermediate set (columns of R / rows of S)
            for (size_t j = 0; j < n; j++) {
                // fuzzy composition rule: max over min
                maxVal = max(maxVal, min(R[i][j], S[j][k]));
            }
            result[i][k] = maxVal; // store best value in result
        }
    }
    return result;
}


void printRelation(const vector<vector<double>>& R) {
    for (const auto& row : R) {  // loop through each row
        cout << "{ ";
        for (double val : row) { // loop through values in row
            cout << val << " ";  // print membership value
        }
        cout << "}" << endl;     // end of row
    }
}

int main() {
    // Define fuzzy relation R (2x3 matrix)
    vector<vector<double>> R = {
        {0.2, 0.7, 1.0},
        {0.5, 0.3, 0.9}
    };

    // Define fuzzy relation S (2x3 matrix)
    vector<vector<double>> S = {
        {0.6, 0.4, 0.8},
        {0.1, 0.9, 0.5}
    };

    // Print the relations
    cout << "Relation R:" << endl; 
    printRelation(R);
    cout << "Relation S:" << endl; 
    printRelation(S);

    // Perform fuzzy union, intersection, and complement
    auto unionRel = fuzzyUnion(R, S);
    auto interRel = fuzzyIntersection(R, S);
    auto compR = fuzzyComplement(R);

    // Print results
    cout << "\nUnion (R ∪ S):" << endl; 
    printRelation(unionRel);

    cout << "\nIntersection (R ∩ S):" << endl; 
    printRelation(interRel);

    cout << "\nComplement (R'):" << endl; 
    printRelation(compR);

    // Define another relation T (3x2 matrix) for composition
    vector<vector<double>> T = {
        {0.3, 0.6},
        {0.8, 0.2},
        {0.4, 0.9}
    };

    cout << "\nRelation T (for composition):" << endl; 
    printRelation(T);

    // Compute composition (R ○ T)
    auto compRel = fuzzyComposition(R, T);
    cout << "\nComposition (R ○ T):" << endl; 
    printRelation(compRel);

    return 0; // program ends successfully
}

File: SC/Assignment2.java
import java.util.ArrayList;
import java.util.Scanner;

public class Assignment2 {

    // Union of two fuzzy relations
    public static ArrayList<ArrayList<Double>> fuzzyUnion(ArrayList<ArrayList<Double>> R,
                                                          ArrayList<ArrayList<Double>> S) {
        ArrayList<ArrayList<Double>> result = new ArrayList<>();

        for (int i = 0; i < R.size(); i++) {
            ArrayList<Double> row = new ArrayList<>();
            for (int j = 0; j < R.get(0).size(); j++) {
                row.add(Math.max(R.get(i).get(j), S.get(i).get(j)));
            }
            result.add(row);
        }
        return result;
    }

    // Intersection of two fuzzy relations
    public static ArrayList<ArrayList<Double>> fuzzyIntersection(ArrayList<ArrayList<Double>> R,
                                                                 ArrayList<ArrayList<Double>> S) {
        ArrayList<ArrayList<Double>> result = new ArrayList<>();

        for (int i = 0; i < R.size(); i++) {
            ArrayList<Double> row = new ArrayList<>();
            for (int j = 0; j < R.get(0).size(); j++) {
                row.add(Math.min(R.get(i).get(j), S.get(i).get(j)));
            }
            result.add(row);
        }

        return result;
    }

    // Complement of a fuzzy relation
    public static ArrayList<ArrayList<Double>> fuzzyComplement(ArrayList<ArrayList<Double>> R) {
        ArrayList<ArrayList<Double>> result = new ArrayList<>();

        for (int i = 0; i < R.size(); i++) {
            ArrayList<Double> row = new ArrayList<>();
            for (int j = 0; j < R.get(0).size(); j++) {
                row.add(1.0 - R.get(i).get(j));
            }
            result.add(row);
        }

        return result;
    }

    // Composition (R ○ T)
    public static ArrayList<ArrayList<Double>> fuzzyComposition(ArrayList<ArrayList<Double>> R,
                                                                ArrayList<ArrayList<Double>> T) {
        int m = R.size();
        int n = R.get(0).size();
        int p = T.get(0).size();

        ArrayList<ArrayList<Double>> result = new ArrayList<>();

        for (int i = 0; i < m; i++) {
            ArrayList<Double> row = new ArrayList<>();

            for (int k = 0; k < p; k++) {
                double maxVal = 0.0;
                for (int j = 0; j < n; j++) {
                    maxVal = Math.max(maxVal, Math.min(R.get(i).get(j), T.get(j).get(k)));
                }
                row.add(maxVal);
            }
            result.add(row);
        }

        return result;
    }

    // Print matrix
    public static void printRelation(ArrayList<ArrayList<Double>> R) {
        for (ArrayList<Double> row : R) {
            System.out.print("{ ");
            for (double val : row) {
                System.out.print(val + " ");
            }
            System.out.println("}");
        }
    }

    // Read matrix input from user
    public static ArrayList<ArrayList<Double>> readRelation(Scanner sc, int rows, int cols, String name) {
        ArrayList<ArrayList<Double>> R = new ArrayList<>();
        System.out.println("Enter relation " + name + " values (" + rows + "x" + cols + "):");

        for (int i = 0; i < rows; i++) {
            ArrayList<Double> row = new ArrayList<>();
            for (int j = 0; j < cols; j++) {
                row.add(sc.nextDouble());
            }
            R.add(row);
        }
        return R;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Input size for R and S
        System.out.print("Enter rows and columns for relations R and S: ");
        int r = sc.nextInt();
        int c = sc.nextInt();

        ArrayList<ArrayList<Double>> R = readRelation(sc, r, c, "R");
        ArrayList<ArrayList<Double>> S = readRelation(sc, r, c, "S");

        System.out.println("\nRelation R:");
        printRelation(R);

        System.out.println("Relation S:");
        printRelation(S);

        var unionRel = fuzzyUnion(R, S);
        var interRel = fuzzyIntersection(R, S);
        var compR = fuzzyComplement(R);

        System.out.println("\nUnion (R ∪ S):");
        printRelation(unionRel);

        System.out.println("\nIntersection (R ∩ S):");
        printRelation(interRel);

        System.out.println("\nComplement (R'):");
        printRelation(compR);

        // Input size for T to perform composition
        System.out.print("\nEnter rows and columns for relation T: ");
        int tR = sc.nextInt();
        int tC = sc.nextInt();

        ArrayList<ArrayList<Double>> T = readRelation(sc, tR, tC, "T");

        System.out.println("\nRelation T:");
        printRelation(T);

        if (c != tR) {
            System.out.println("\nComposition not possible: Columns of R must equal rows of T");
        } else {
            var compRel = fuzzyComposition(R, T);
            System.out.println("\nComposition (R ○ T):");
            printRelation(compRel);
        }

        sc.close();
    }
}
